modules/network/src/main/java/org/apache/ignite/network/internal/netty/NettyServer.java

```diff
@@ -0,0 +1,254 @@
+        this.newConnectionListener = newConnectionListener;
+        this.messageListener = messageListener;
+        this.serializationRegistry = serializationRegistry;
+    }
+
+    /**
+     * Starts the server.
+     *
+     * @return Future that resolves when the server is successfully started.
+     */
+    public CompletableFuture<Void> start() {
+        if (serverStartFuture != null)
+            throw new IgniteInternalException("Attempted to start an already started server");
+
+        bootstrap.group(bossGroup, workerGroup)
+            .channel(NioServerSocketChannel.class)
+            .childHandler(new ChannelInitializer<SocketChannel>() {
+                /** {@inheritDoc} */
+                @Override public void initChannel(SocketChannel ch) {
+                    ch.pipeline().addLast(
+                        /*
+                         * Decoder that uses org.apache.ignite.network.internal.MessageReader
+                         * to read chunked data.
+                         */
+                        new InboundDecoder(serializationRegistry),
+                        // Handles decoded NetworkMessages.
+                        new MessageHandler(messageListener),
+                        /*
+                         * Encoder that uses org.apache.ignite.network.internal.MessageWriter
+                         * to write chunked data.
+                         */
+                        new ChunkedWriteHandler()
+                    );
+
+                    newConnectionListener.accept(new NettySender(ch, serializationRegistry));
+                }
+            })
+            /*
+             * The maximum queue length for incoming connection indications (a request to connect) is set
+             * to the backlog parameter. If a connection indication arrives when the queue is full,
+             * the connection is refused.
+             */
+            .option(ChannelOption.SO_BACKLOG, 128)
+            /*
+             * When the keepalive option is set for a TCP socket and no data has been exchanged across the socket
+             * in either direction for 2 hours (NOTE: the actual value is implementation dependent),
+             * TCP automatically sends a keepalive probe to the peer.
+             */
+            .childOption(ChannelOption.SO_KEEPALIVE, true);
+
+        serverStartFuture = new CompletableFuture<>();
+
+        NettyUtils.toChannelCompletableFuture(bootstrap.bind(port))
+            .whenComplete((channel, err) -> {
+                synchronized (this) {
+                    // Shutdown event loops if the server has failed to start of has been stopped.
+                    if (err != null || stopped) {
+                        this.channel = null;
+
+                        shutdownEventLoopGroups();
+
+                        serverCloseFuture.whenComplete((unused, throwable) -> {
+                            Throwable stopErr = err != null ? err : new CancellationException();
```

Comment:
1. Why `CancellationException`?
2. Add a message to the exception that the server has been stopped.

Reply:
1. I think CancellationException is suitable here, as we cancelled server's startup.
2. Ok